library/bignum_mod.h

```diff
@@ -0,0 +1,142 @@
+/**
+ *  Internal bignum functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef MBEDTLS_BIGNUM_MOD_H
+#define MBEDTLS_BIGNUM_MOD_H
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+#include "mbedtls/bignum.h"
+#endif
+
+
+typedef struct
+{
+    size_t n;
+    mbedtls_mpi_uint *p;
+} mbedtls_mpi_mod_residue;
+
+typedef void* mbedtls_mpi_mont_struct;
+typedef void* mbedtls_mpi_opt_red_struct;
+
+typedef struct {
+    mbedtls_mpi_uint *p;
+    size_t n;       // number of limbs
+    size_t plen;    // bitlen of p
+    int ext_rep;    // signals external representation (eg. byte order)
+    int int_rep;    // selector to signal the active member of the union
+    union rep
+    {
+        mbedtls_mpi_mont_struct mont;
+        mbedtls_mpi_opt_red_struct ored;
+    } rep;
+} mbedtls_mpi_mod_modulus;
+
+typedef enum
```

Comment:
the `enum` definitions should come before `mbedtls_mpi_mod_modulus` and be used within it, surely?

Reply:
It can be before it and eventually it could be used within it.

The reason why they are `int` in the struct and in function signatures is that the exact type of enums is implementation defined. If the compiler that compiled the shared object library has different ideas about enums, that breaks binary compatibility. Of course exactly because of this most compilers will have it as an `int`, but still, it is more portable this way.

In the end, these will be only used inside libmbedcrypto, but there will be a transition period where these might be used from X509 or TLS. Again, I can see that compiling these with different compilers might not be the most common use case, but was trying to go for portability.

We might be able to get away with doing it now, but it seemed safer this way.

Reply:
Understood

Reply:
Actually, the more I think about it the more I think that we should use enum definitions in structs and functions:

- This is an internal API and we could get out a lot of the compiler warnings (we would use compilers that have warnings and would actually fix them)
- If we remove the invalid value from the enum, this would allow us to remove enum checks and save code size
- Having the three libraries compiled with different compiler seems to be a very unlikely and niche use case

What do you think?

Reply:
I would be very happy for us to do that. ("Understood" = "it seems like you have reasons, although this wouldn't be how I would do it 'in a vacuum'")

Regarding size of the field, compilers will typically either just use `int` or use the smallest "integer" size that fits all the values, so in this case `char`. I can't see us having large numbers of different values in the enums (I would be surprised if there were more than 10 or 20) then as the enum changes the size of field chosen is unlikely to change.

(A common idiom when it's expected to have significant increase in the number of enum values is to define a "max"in the enum, but I don't think that's necessary here.)

And this would also address one of my other comments, about inconsistency in the PR of sometimes using `switch` and sometimes `if/else` - we should always use `switch`, because there are compiler switches to catch it when a switch doesn't handle every enum value.

i.e. YES PLEASE!

Reply:
We're already using enums in public structs. E.g. there are several in `mbedtls_cipher_info_t`. So you already need to build libmbedcrypto and your application with compatible ABIs.